### DISCUSSION TITLE
Exercise 3.3 - do we need to use nextTick() ?

### DISCUSSION THREAD

**Discussion by stoptime** [PacktPublishing/Node.js-Design-Patterns-Third-Edition#31] [Q&A] [answered]
Hi all - really enjoying this book!

I was wondering if in exercise 3.3 if we need to use the `process.nextTick()` function to emit a 'tick' immediately after the function is invoked? I notice that if I do not, and simply do something like `emitter.emit('tick', '3.3 update')` that it never fires at all, even if I remove the loop, and I'm trying to wrap my head around that... 🤔

```
import { EventEmitter } from 'events'
import { nextTick } from 'process'

function everyFifty(totalMil, cb) {
  let startMil = 0
  let ticks = 0
  const emitter = new EventEmitter()
  nextTick(() => emitter.emit('tick', '3.3 update')) // this will not fire w/o nextTick()

  while (startMil <= totalMil) {
    startMil += 50
    if (startMil > totalMil) {
      setTimeout(() => {
        cb(ticks)
      }, totalMil)
      break
    }
    setTimeout(() => {
      emitter.emit('tick', (ticks += 1))
    }, startMil)
  }

  return emitter
}

function finished(total) {
  console.log('completed total ticks:', total)
}

everyFifty(100, finished).on('tick', (ticks) => {
  console.log('tick', ticks)
})

// -- output --
// tick 3.3 update
// tick 1
// tick 2
// completed total ticks: 2

```

**Comment by mariocasciaro** [comment, marked as answer]
Hi Matt, thanks for reading the book. Yes you'll need a nextTick. That's
essentially the main takeaway from exercise 3.3 😉. Without it, the event
is emitted before the event emitter is even returned to the caller. The
event fires but it's not intercepted by any listener, because there are
none at the point.

**Reply by stoptime** [reply]
Ah-ha! 💡 I think there was just an audible 'click' emitted from my head - thank you!
